{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(file_path):\n",
    "    \"\"\"\n",
    "    Reads a FASTA file and returns a dictionary where keys are the headers (chromosomes or enzyme names)\n",
    "    and values are the corresponding sequences.\n",
    "    \"\"\"\n",
    "    with open(file_path, 'r') as file:\n",
    "        sequences = {}\n",
    "        header = None # Initiating header to None\n",
    "        sequence = []\n",
    "\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if header:\n",
    "                    sequences[header] = ''.join(sequence)\n",
    "                header = line[1:]  # Remove '>' character\n",
    "                sequence = []\n",
    "            else:\n",
    "                sequence.append(line)\n",
    "        \n",
    "        # Add the last sequence\n",
    "        if header:\n",
    "            sequences[header] = ''.join(sequence)\n",
    "\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def restriction_sites(file_path):\n",
    "    \"\"\"\n",
    "    Reads the enzyme file and returns a dictionary where keys are the enzymes and values are a list\n",
    "    consisting cutsite and the nick.\n",
    "    \"\"\"\n",
    "    enzymes = read_fasta(file_path)\n",
    "    for key in enzymes:\n",
    "        nick = enzymes[key].find(\"/\")\n",
    "        enzymes[key] = [enzymes[key].replace(\"/\", \"\").replace(\"N\", \"[ACTG]\").replace(\"Y\", \"[CT]\").replace(\"R\", \"[AG]\"), nick]\n",
    "        # Replace all the placeholders with the corresponding regex\n",
    "    return enzymes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def find_restriction_sites(genome, enzyme):\n",
    "    \"\"\"\n",
    "    Finds restriction sites in the genome. Returns a dictionary where keys are headers (chromosomes)\n",
    "    and values are lists of positions of the restriction site.\n",
    "    \"\"\"\n",
    "    positions = {}\n",
    "    restriction_site = restriction_sites(\"test_enzyme.fa\")[enzyme]\n",
    "\n",
    "    for chr, seq in genome.items():\n",
    "        positions[chr] = [0]+ [restriction_site[1] + m.start() for m in re.finditer(restriction_site[0], seq)]\n",
    "        # Find all the restriction sites and add the nick to the position, add the first position of the chromosome\n",
    "        if positions[chr][-1] != len(seq):\n",
    "            positions[chr].append(len(seq))\n",
    "        # Add the last position of the chromosome\n",
    "\n",
    "    return positions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def random_fragment(sites, min_size, max_size):\n",
    "    \"\"\"\n",
    "    Input is a list of restriction sites, and the maximum and minimum size of the fragment,\n",
    "    output is a list of restriction sites inserted with random break sites.\n",
    "    \"\"\"\n",
    "    import random\n",
    "    new_sites = sites.copy()  # Copy the original list to avoid modifying it\n",
    "    current_position = sites[0]\n",
    "\n",
    "    while current_position < sites[-1]:\n",
    "        current_position += random.randint(min_size, max_size)\n",
    "        if current_position < sites[-1]:\n",
    "            new_sites.append(current_position)\n",
    "\n",
    "    return sorted(new_sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NM_000245.3 Homo sapiens MET proto-oncogene, receptor tyrosine kinase (MET), transcript variant 2, mRNA': 'GCAGGTGACCCGGAGGCCCTCGCCGCCCGCGGCGCCCCGAGCGCTTTGTGAGCAGATGCGGAGCCGAGTGGAGGGCGCGAGCCAGATGCGGGGCGACAGCTGACTTGCTGAGAGGAGGCGGGGAGGCGCGGAGCGCGCGTGTGGTCCTTGCGCCGCTGACTTCTCCACTGGTTCCTGGGCACCGAAAGATAAACCTCTCATAATGAAGGCCCCCGCTGTGCTTGCACCTGGCATCCTCGTGCTCCTGTTTACCTTGGTGCAGAGGAGCAATGGGGAGTGTAAAGAGGCACTAGCAAAGTCCGAGATGAATGTGAATATGAAGTATCAGCTTCCCAACTTCACCGCGGAAACACCCATCCAGAATGTCATTCTACATGAGCATCACATTTTCCTTGGTGCCACTAACTACATTTATGTTTTAAATGAGGAAGACCTTCAGAAGGTTGCTGAGTACAAGACTGGGCCTGTGCTGGAACACCCAGATTGTTTCCCATGTCAGGACTGCAGCAGCAAAGCCAATTTATCAGGAGGTGTTTGGAAAGATAACATCAACATGGCTCTAGTTGTCGACACCTACTATGATGATCAACTCATTAGCTGTGGCAGCGTCAACAGAGGGACCTGCCAGCGACATGTCTTTCCCCACAATCATACTGCTGACATACAGTCGGAGGTTCACTGCATATTCTCCCCACAGATAGAAGAGCCCAGCCAGTGTCCTGACTGTGTGGTGAGCGCCCTGGGAGCCAAAGTCCTTTCATCTGTAAAGGACCGGTTCATCAACTTCTTTGTAGGCAATACCATAAATTCTTCTTATTTCCCAGATCATCCATTGCATTCGATATCAGTGAGAAGGCTAAAGGAAACGAAAGATGGTTTTATGTTTTTGACGGACCAGTCCTACATTGATGTTTTACCTGAGTTCAGAGATTCTTACCCCATTAAGTATGTCCATGCCTTTGAAAGCAACAATTTTATTTACTTCTTGACGGTCCAAAGGGAAACTCTAGATGCTCAGACTTTTCACACAAGAATAATCAGGTTCTGTTCCATAAACTCTGGATTGCATTCCTACATGGAAATGCCTCTGGAGTGTATTCTCACAGAAAAGAGAAAAAAGAGATCCACAAAGAAGGAAGTGTTTAATATACTTCAGGCTGCGTATGTCAGCAAGCCTGGGGCCCAGCTTGCTAGACAAATAGGAGCCAGCCTGAATGATGACATTCTTTTCGGGGTGTTCGCACAAAGCAAGCCAGATTCTGCCGAACCAATGGATCGATCTGCCATGTGTGCATTCCCTATCAAATATGTCAACGACTTCTTCAACAAGATCGTCAACAAAAACAATGTGAGATGTCTCCAGCATTTTTACGGACCCAATCATGAGCACTGCTTTAATAGGACACTTCTGAGAAATTCATCAGGCTGTGAAGCGCGCCGTGATGAATATCGAACAGAGTTTACCACAGCTTTGCAGCGCGTTGACTTATTCATGGGTCAATTCAGCGAAGTCCTCTTAACATCTATATCCACCTTCATTAAAGGAGACCTCACCATAGCTAATCTTGGGACATCAGAGGGTCGCTTCATGCAGGTTGTGGTTTCTCGATCAGGACCATCAACCCCTCATGTGAATTTTCTCCTGGACTCCCATCCAGTGTCTCCAGAAGTGATTGTGGAGCATACATTAAACCAAAATGGCTACACACTGGTTATCACTGGGAAGAAGATCACGAAGATCCCATTGAATGGCTTGGGCTGCAGACATTTCCAGTCCTGCAGTCAATGCCTCTCTGCCCCACCCTTTGTTCAGTGTGGCTGGTGCCACGACAAATGTGTGCGATCGGAGGAATGCCTGAGCGGGACATGGACTCAACAGATCTGTCTGCCTGCAATCTACAAGGTTTTCCCAAATAGTGCACCCCTTGAAGGAGGGACAAGGCTGACCATATGTGGCTGGGACTTTGGATTTCGGAGGAATAATAAATTTGATTTAAAGAAAACTAGAGTTCTCCTTGGAAATGAGAGCTGCACCTTGACTTTAAGTGAGAGCACGATGAATACATTGAAATGCACAGTTGGTCCTGCCATGAATAAGCATTTCAATATGTCCATAATTATTTCAAATGGCCACGGGACAACACAATACAGTACATTCTCCTATGTGGATCCTGTAATAACAAGTATTTCGCCGAAATACGGTCCTATGGCTGGTGGCACTTTACTTACTTTAACTGGAAATTACCTAAACAGTGGGAATTCTAGACACATTTCAATTGGTGGAAAAACATGTACTTTAAAAAGTGTGTCAAACAGTATTCTTGAATGTTATACCCCAGCCCAAACCATTTCAACTGAGTTTGCTGTTAAATTGAAAATTGACTTAGCCAACCGAGAGACAAGCATCTTCAGTTACCGTGAAGATCCCATTGTCTATGAAATTCATCCAACCAAATCTTTTATTAGTGGTGGGAGCACAATAACAGGTGTTGGGAAAAACCTGAATTCAGTTAGTGTCCCGAGAATGGTCATAAATGTGCATGAAGCAGGAAGGAACTTTACAGTGGCATGTCAACATCGCTCTAATTCAGAGATAATCTGTTGTACCACTCCTTCCCTGCAACAGCTGAATCTGCAACTCCCCCTGAAAACCAAAGCCTTTTTCATGTTAGATGGGATCCTTTCCAAATACTTTGATCTCATTTATGTACATAATCCTGTGTTTAAGCCTTTTGAAAAGCCAGTGATGATCTCAATGGGCAATGAAAATGTACTGGAAATTAAGGGAAATGATATTGACCCTGAAGCAGTTAAAGGTGAAGTGTTAAAAGTTGGAAATAAGAGCTGTGAGAATATACACTTACATTCTGAAGCCGTTTTATGCACGGTCCCCAATGACCTGCTGAAATTGAACAGCGAGCTAAATATAGAGTGGAAGCAAGCAATTTCTTCAACCGTCCTTGGAAAAGTAATAGTTCAACCAGATCAGAATTTCACAGGATTGATTGCTGGTGTTGTCTCAATATCAACAGCACTGTTATTACTACTTGGGTTTTTCCTGTGGCTGAAAAAGAGAAAGCAAATTAAAGATCTGGGCAGTGAATTAGTTCGCTACGATGCAAGAGTACACACTCCTCATTTGGATAGGCTTGTAAGTGCCCGAAGTGTAAGCCCAACTACAGAAATGGTTTCAAATGAATCTGTAGACTACCGAGCTACTTTTCCAGAAGATCAGTTTCCTAATTCATCTCAGAACGGTTCATGCCGACAAGTGCAGTATCCTCTGACAGACATGTCCCCCATCCTAACTAGTGGGGACTCTGATATATCCAGTCCATTACTGCAAAATACTGTCCACATTGACCTCAGTGCTCTAAATCCAGAGCTGGTCCAGGCAGTGCAGCATGTAGTGATTGGGCCCAGTAGCCTGATTGTGCATTTCAATGAAGTCATAGGAAGAGGGCATTTTGGTTGTGTATATCATGGGACTTTGTTGGACAATGATGGCAAGAAAATTCACTGTGCTGTGAAATCCTTGAACAGAATCACTGACATAGGAGAAGTTTCCCAATTTCTGACCGAGGGAATCATCATGAAAGATTTTAGTCATCCCAATGTCCTCTCGCTCCTGGGAATCTGCCTGCGAAGTGAAGGGTCTCCGCTGGTGGTCCTACCATACATGAAACATGGAGATCTTCGAAATTTCATTCGAAATGAGACTCATAATCCAACTGTAAAAGATCTTATTGGCTTTGGTCTTCAAGTAGCCAAAGGCATGAAATATCTTGCAAGCAAAAAGTTTGTCCACAGAGACTTGGCTGCAAGAAACTGTATGCTGGATGAAAAATTCACAGTCAAGGTTGCTGATTTTGGTCTTGCCAGAGACATGTATGATAAAGAATACTATAGTGTACACAACAAAACAGGTGCAAAGCTGCCAGTGAAGTGGATGGCTTTGGAAAGTCTGCAAACTCAAAAGTTTACCACCAAGTCAGATGTGTGGTCCTTTGGCGTGCTCCTCTGGGAGCTGATGACAAGAGGAGCCCCACCTTATCCTGACGTAAACACCTTTGATATAACTGTTTACTTGTTGCAAGGGAGAAGACTCCTACAACCCGAATACTGCCCAGACCCCTTATATGAAGTAATGCTAAAATGCTGGCACCCTAAAGCCGAAATGCGCCCATCCTTTTCTGAACTGGTGTCCCGGATATCAGCGATCTTCTCTACTTTCATTGGGGAGCACTATGTCCATGTGAACGCTACTTATGTGAACGTAAAATGTGTCGCTCCGTATCCTTCTCTGTTGTCATCAGAAGATAACGCTGATGATGAGGTGGACACACGACCAGCCTCCTTCTGGGAGACATCATAGTGCTAGTACTATGTCAAAGCAACAGTCCACACTTTGTCCAATGGTTTTTTCACTGCCTGACCTTTAAAAGGCCATCGATATTCTTTGCTCTTGCCAAAATTGCACTATTATAGGACTTGTATTGTTATTTAAATTACTGGATTCTAAGGAATTTCTTATCTGACAGAGCATCAGAACCAGAGGCTTGGTCCCACAGGCCACGGACCAATGGCCTGCAGCCGTGACAACACTCCTGTCATATTGGAGTCCAAAACTTGAATTCTGGGTTGAATTTTTTAAAAATCAGGTACCACTTGATTTCATATGGGAAATTGAAGCAGGAAATATTGAGGGCTTCTTGATCACAGAAAACTCAGAAGAGATAGTAATGCTCAGGACAGGAGCGGCAGCCCCAGAACAGGCCACTCATTTAGAATTCTAGTGTTTCAAAACACTTTTGTGTGTTGTATGGTCAATAACATTTTTCATTACTGATGGTGTCATTCACCCATTAGGTAAACATTCCCTTTTAAATGTTTGTTTGTTTTTTGAGACAGGATCTCACTCTGTTGCCAGGGCTGTAGTGCAGTGGTGTGATCATAGCTCACTGCAACCTCCACCTCCCAGGCTCAAGCCTCCCGAATAGCTGGGACTACAGGCGCACACCACCATCCCCGGCTAATTTTTGTATTTTTTGTAGAGACGGGGTTTTGCCATGTTGCCAAGGCTGGTTTCAAACTCCTGGACTCAAGAAATCCACCCACCTCAGCCTCCCAAAGTGCTAGGATTACAGGCATGAGCCACTGCGCCCAGCCCTTATAAATTTTTGTATAGACATTCCTTTGGTTGGAAGAATATTTATAGGCAATACAGTCAAAGTTTCAAAATAGCATCACACAAAACATGTTTATAAATGAACAGGATGTAATGTACATAGATGACATTAAGAAAATTTGTATGAAATAATTTAGTCATCATGAAATATTTAGTTGTCATATAAAAACCCACTGTTTGAGAATGATGCTACTCTGATCTAATGAATGTGAACATGTAGATGTTTTGTGTGTATTTTTTTAAATGAAAACTCAAAATAAGACAAGTAATTTGTTGATAAATATTTTTAAAGATAACTCAGCATGTTTGTAAAGCAGGATACATTTTACTAAAAGGTTCATTGGTTCCAATCACAGCTCATAGGTAGAGCAAAGAAAGGGTGGATGGATTGAAAAGATTAGCCTCTGTCTCGGTGGCAGGTTCCCACCTCGCAAGCAATTGGAAACAAAACTTTTGGGGAGTTTTATTTTGCATTAGGGTGTGTTTTATGTTAAGCAAAACATACTTTAGAAACAAATGAAAAAGGCAATTGAAAATCCCAGCTATTTCACCTAGATGGAATAGCCACCCTGAGCAGAACTTTGTGATGCTTCATTCTGTGGAATTTTGTGCTTGCTACTGTATAGTGCATGTGGTGTAGGTTACTCTAACTGGTTTTGTCGACGTAAACATTTAAAGTGTTATATTTTTTATAAAAATGTTTATTTTTAATGATATGAGAAAAATTTTGTTAGGCCACAAAAACACTGCACTGTGAACATTTTAGAAAAGGTATGTCAGACTGGGATTAATGACAGCATGATTTTCAATGACTGTAAATTGCGATAAGGAAATGTACTGATTGCCAATACACCCCACCCTCATTACATCATCAGGACTTGAAGCCAAGGGTTAACCCAGCAAGCTACAAAGAGGGTGTGTCACACTGAAACTCAATAGTTGAGTTTGGCTGTTGTTGCAGGAAAATGATTATAACTAAAAGCTCTCTGATAGTGCAGAGACTTACCAGAAGACACAAGGAATTGTACTGAAGAGCTATTACAATCCAAATATTGCCGTTTCATAAATGTAATAAGTAATACTAATTCACAGAGTATTGTAAATGGTGGATGACAAAAGAAAATCTGCTCTGTGGAAAGAAAGAACTGTCTCTACCAGGGTCAAGAGCATGAACGCATCAATAGAAAGAACTCGGGGAAACATCCCATCAACAGGACTACACACTTGTATATACATTCTTGAGAACACTGCAATGTGAAAATCACGTTTGCTATTTATAAACTTGTCCTTAGATTAATGTGTCTGGACAGATTGTGGGAGTAAGTGATTCTTCTAAGAATTAGATACTTGTCACTGCCTATACCTGCAGCTGAACTGAATGGTACTTCGTATGTTAATAGTTGTTCTGATAAATCATGCAATTAAAGTAAAGTGATGCAACATCTTGTAAAAAAAAAAAAAAAAAAAA'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def fetch_fasta_from_ncbi(accession_number):\n",
    "    \"\"\"\n",
    "    Fetches the fasta file from NCBI using the accession number.\n",
    "    \"\"\"\n",
    "    # Please run \"Install Certificates.command\" in your Python directory first if you are using MacOS\n",
    "    import requests\n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "    params = {\n",
    "        'db': 'nucleotide',\n",
    "        'id': accession_number,\n",
    "        'rettype': 'fasta',\n",
    "        'retmode': 'text'\n",
    "    }\n",
    "    response = requests.get(base_url, params=params)\n",
    "    if response.status_code != 200:\n",
    "        raise Exception(f\"Error fetching data from NCBI: HTTP Status {response.status_code}\")\n",
    "\n",
    "    # Split the record into lines and filter out empty lines\n",
    "    record_lines = [line for line in response.text.strip().split(\"\\n\") if line]\n",
    "    \n",
    "    # Reformat the record into a dictionary with key:value as header:sequence, if there is more than one header exists in the record that starts with \">\"\n",
    "    sequences = {}\n",
    "    for line in record_lines:\n",
    "        if line.startswith(\">\"):\n",
    "            header = line[1:]\n",
    "            if header not in sequences:\n",
    "                sequences[header] = \"\"\n",
    "        else:\n",
    "            sequences[header] += line\n",
    "    \n",
    "    return sequences\n",
    "\n",
    "fetch_fasta_from_ncbi(\"NM_000245.3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb Cell 6\u001b[0m line \u001b[0;36m1\n\u001b[1;32m    <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=114'>115</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m distances \u001b[39m# Return the distances list to be used in the plot\u001b[39;00m\n\u001b[1;32m    <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=117'>118</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39m__name__\u001b[39m \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m__main__\u001b[39m\u001b[39m\"\u001b[39m:\n\u001b[0;32m--> <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=118'>119</a>\u001b[0m     main()\n",
      "\u001b[1;32m/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb Cell 6\u001b[0m line \u001b[0;36m1\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m enzyme_file \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mtest_enzyme.fa\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m \u001b[39m# Read genome and restriction enzyme sites\u001b[39;00m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m genome \u001b[39m=\u001b[39m fetch_fasta_from_ncbi(\u001b[39m\"\u001b[39;49m\u001b[39mNC_037328.1\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m enzymes \u001b[39m=\u001b[39m restriction_sites(enzyme_file)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m \u001b[39m# Create a dictionary to store the restriction sites for the selected enzymes\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=13'>14</a>\u001b[0m \u001b[39m# selected_enzymes = [enzyme.strip() for enzyme in input(\"Enter the names of the enzymes separated by commas (e.g., EcoRI, HindIII): \").split(',')]\u001b[39;00m\n",
      "\u001b[1;32m/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb Cell 6\u001b[0m line \u001b[0;36m1\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m base_url \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mhttps://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m params \u001b[39m=\u001b[39m {\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mdb\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mnucleotide\u001b[39m\u001b[39m'\u001b[39m,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mid\u001b[39m\u001b[39m'\u001b[39m: accession_number,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mrettype\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mfasta\u001b[39m\u001b[39m'\u001b[39m,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mretmode\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mtext\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m }\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m response \u001b[39m=\u001b[39m requests\u001b[39m.\u001b[39;49mget(base_url, params\u001b[39m=\u001b[39;49mparams)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=16'>17</a>\u001b[0m \u001b[39mif\u001b[39;00m response\u001b[39m.\u001b[39mstatus_code \u001b[39m!=\u001b[39m \u001b[39m200\u001b[39m:\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/jeromexqian/genomedigester/test/test_from_NCBI.ipynb#W5sZmlsZQ%3D%3D?line=17'>18</a>\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mException\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mError fetching data from NCBI: HTTP Status \u001b[39m\u001b[39m{\u001b[39;00mresponse\u001b[39m.\u001b[39mstatus_code\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget\u001b[39m(url, params\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m \u001b[39m    \u001b[39m\u001b[39mr\u001b[39m\u001b[39m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[39m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[39m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[39mreturn\u001b[39;00m request(\u001b[39m\"\u001b[39;49m\u001b[39mget\u001b[39;49m\u001b[39m\"\u001b[39;49m, url, params\u001b[39m=\u001b[39;49mparams, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[39m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[39m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[39m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[39mwith\u001b[39;00m sessions\u001b[39m.\u001b[39mSession() \u001b[39mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[39mreturn\u001b[39;00m session\u001b[39m.\u001b[39;49mrequest(method\u001b[39m=\u001b[39;49mmethod, url\u001b[39m=\u001b[39;49murl, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[39m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mtimeout\u001b[39m\u001b[39m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mallow_redirects\u001b[39m\u001b[39m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[39m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msend(prep, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49msend_kwargs)\n\u001b[1;32m    591\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/requests/sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m start \u001b[39m=\u001b[39m preferred_clock()\n\u001b[1;32m    702\u001b[0m \u001b[39m# Send the request\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m r \u001b[39m=\u001b[39m adapter\u001b[39m.\u001b[39;49msend(request, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    705\u001b[0m \u001b[39m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    706\u001b[0m elapsed \u001b[39m=\u001b[39m preferred_clock() \u001b[39m-\u001b[39m start\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/requests/adapters.py:486\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    483\u001b[0m     timeout \u001b[39m=\u001b[39m TimeoutSauce(connect\u001b[39m=\u001b[39mtimeout, read\u001b[39m=\u001b[39mtimeout)\n\u001b[1;32m    485\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 486\u001b[0m     resp \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49murlopen(\n\u001b[1;32m    487\u001b[0m         method\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mmethod,\n\u001b[1;32m    488\u001b[0m         url\u001b[39m=\u001b[39;49murl,\n\u001b[1;32m    489\u001b[0m         body\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mbody,\n\u001b[1;32m    490\u001b[0m         headers\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mheaders,\n\u001b[1;32m    491\u001b[0m         redirect\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    492\u001b[0m         assert_same_host\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    493\u001b[0m         preload_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    494\u001b[0m         decode_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    495\u001b[0m         retries\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mmax_retries,\n\u001b[1;32m    496\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[1;32m    497\u001b[0m         chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    498\u001b[0m     )\n\u001b[1;32m    500\u001b[0m \u001b[39mexcept\u001b[39;00m (ProtocolError, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m    501\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m(err, request\u001b[39m=\u001b[39mrequest)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/urllib3/connectionpool.py:790\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, preload_content, decode_content, **response_kw)\u001b[0m\n\u001b[1;32m    787\u001b[0m response_conn \u001b[39m=\u001b[39m conn \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m release_conn \u001b[39melse\u001b[39;00m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m    789\u001b[0m \u001b[39m# Make the request on the HTTPConnection object\u001b[39;00m\n\u001b[0;32m--> 790\u001b[0m response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    791\u001b[0m     conn,\n\u001b[1;32m    792\u001b[0m     method,\n\u001b[1;32m    793\u001b[0m     url,\n\u001b[1;32m    794\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    795\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    796\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    797\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    798\u001b[0m     retries\u001b[39m=\u001b[39;49mretries,\n\u001b[1;32m    799\u001b[0m     response_conn\u001b[39m=\u001b[39;49mresponse_conn,\n\u001b[1;32m    800\u001b[0m     preload_content\u001b[39m=\u001b[39;49mpreload_content,\n\u001b[1;32m    801\u001b[0m     decode_content\u001b[39m=\u001b[39;49mdecode_content,\n\u001b[1;32m    802\u001b[0m     \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mresponse_kw,\n\u001b[1;32m    803\u001b[0m )\n\u001b[1;32m    805\u001b[0m \u001b[39m# Everything went great!\u001b[39;00m\n\u001b[1;32m    806\u001b[0m clean_exit \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/urllib3/connectionpool.py:467\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)\u001b[0m\n\u001b[1;32m    464\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    465\u001b[0m     \u001b[39m# Trigger any extra validation we need to do.\u001b[39;00m\n\u001b[1;32m    466\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 467\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_validate_conn(conn)\n\u001b[1;32m    468\u001b[0m     \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    469\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_raise_timeout(err\u001b[39m=\u001b[39me, url\u001b[39m=\u001b[39murl, timeout_value\u001b[39m=\u001b[39mconn\u001b[39m.\u001b[39mtimeout)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/urllib3/connectionpool.py:1092\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   1090\u001b[0m \u001b[39m# Force connect early to allow us to validate the connection.\u001b[39;00m\n\u001b[1;32m   1091\u001b[0m \u001b[39mif\u001b[39;00m conn\u001b[39m.\u001b[39mis_closed:\n\u001b[0;32m-> 1092\u001b[0m     conn\u001b[39m.\u001b[39;49mconnect()\n\u001b[1;32m   1094\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m conn\u001b[39m.\u001b[39mis_verified:\n\u001b[1;32m   1095\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\n\u001b[1;32m   1096\u001b[0m         (\n\u001b[1;32m   1097\u001b[0m             \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mUnverified HTTPS request is being made to host \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mconn\u001b[39m.\u001b[39mhost\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m. \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1102\u001b[0m         InsecureRequestWarning,\n\u001b[1;32m   1103\u001b[0m     )\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/urllib3/connection.py:642\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    633\u001b[0m \u001b[39mif\u001b[39;00m is_time_off:\n\u001b[1;32m    634\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\n\u001b[1;32m    635\u001b[0m         (\n\u001b[1;32m    636\u001b[0m             \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mSystem time is way off (before \u001b[39m\u001b[39m{\u001b[39;00mRECENT_DATE\u001b[39m}\u001b[39;00m\u001b[39m). This will probably \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    639\u001b[0m         SystemTimeWarning,\n\u001b[1;32m    640\u001b[0m     )\n\u001b[0;32m--> 642\u001b[0m sock_and_verified \u001b[39m=\u001b[39m _ssl_wrap_socket_and_match_hostname(\n\u001b[1;32m    643\u001b[0m     sock\u001b[39m=\u001b[39;49msock,\n\u001b[1;32m    644\u001b[0m     cert_reqs\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcert_reqs,\n\u001b[1;32m    645\u001b[0m     ssl_version\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mssl_version,\n\u001b[1;32m    646\u001b[0m     ssl_minimum_version\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mssl_minimum_version,\n\u001b[1;32m    647\u001b[0m     ssl_maximum_version\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mssl_maximum_version,\n\u001b[1;32m    648\u001b[0m     ca_certs\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mca_certs,\n\u001b[1;32m    649\u001b[0m     ca_cert_dir\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mca_cert_dir,\n\u001b[1;32m    650\u001b[0m     ca_cert_data\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mca_cert_data,\n\u001b[1;32m    651\u001b[0m     cert_file\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcert_file,\n\u001b[1;32m    652\u001b[0m     key_file\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mkey_file,\n\u001b[1;32m    653\u001b[0m     key_password\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mkey_password,\n\u001b[1;32m    654\u001b[0m     server_hostname\u001b[39m=\u001b[39;49mserver_hostname,\n\u001b[1;32m    655\u001b[0m     ssl_context\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mssl_context,\n\u001b[1;32m    656\u001b[0m     tls_in_tls\u001b[39m=\u001b[39;49mtls_in_tls,\n\u001b[1;32m    657\u001b[0m     assert_hostname\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49massert_hostname,\n\u001b[1;32m    658\u001b[0m     assert_fingerprint\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49massert_fingerprint,\n\u001b[1;32m    659\u001b[0m )\n\u001b[1;32m    660\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msock \u001b[39m=\u001b[39m sock_and_verified\u001b[39m.\u001b[39msocket\n\u001b[1;32m    661\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mis_verified \u001b[39m=\u001b[39m sock_and_verified\u001b[39m.\u001b[39mis_verified\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/urllib3/connection.py:783\u001b[0m, in \u001b[0;36m_ssl_wrap_socket_and_match_hostname\u001b[0;34m(sock, cert_reqs, ssl_version, ssl_minimum_version, ssl_maximum_version, cert_file, key_file, key_password, ca_certs, ca_cert_dir, ca_cert_data, assert_hostname, assert_fingerprint, server_hostname, ssl_context, tls_in_tls)\u001b[0m\n\u001b[1;32m    780\u001b[0m     \u001b[39mif\u001b[39;00m is_ipaddress(normalized):\n\u001b[1;32m    781\u001b[0m         server_hostname \u001b[39m=\u001b[39m normalized\n\u001b[0;32m--> 783\u001b[0m ssl_sock \u001b[39m=\u001b[39m ssl_wrap_socket(\n\u001b[1;32m    784\u001b[0m     sock\u001b[39m=\u001b[39;49msock,\n\u001b[1;32m    785\u001b[0m     keyfile\u001b[39m=\u001b[39;49mkey_file,\n\u001b[1;32m    786\u001b[0m     certfile\u001b[39m=\u001b[39;49mcert_file,\n\u001b[1;32m    787\u001b[0m     key_password\u001b[39m=\u001b[39;49mkey_password,\n\u001b[1;32m    788\u001b[0m     ca_certs\u001b[39m=\u001b[39;49mca_certs,\n\u001b[1;32m    789\u001b[0m     ca_cert_dir\u001b[39m=\u001b[39;49mca_cert_dir,\n\u001b[1;32m    790\u001b[0m     ca_cert_data\u001b[39m=\u001b[39;49mca_cert_data,\n\u001b[1;32m    791\u001b[0m     server_hostname\u001b[39m=\u001b[39;49mserver_hostname,\n\u001b[1;32m    792\u001b[0m     ssl_context\u001b[39m=\u001b[39;49mcontext,\n\u001b[1;32m    793\u001b[0m     tls_in_tls\u001b[39m=\u001b[39;49mtls_in_tls,\n\u001b[1;32m    794\u001b[0m )\n\u001b[1;32m    796\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    797\u001b[0m     \u001b[39mif\u001b[39;00m assert_fingerprint:\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/urllib3/util/ssl_.py:444\u001b[0m, in \u001b[0;36mssl_wrap_socket\u001b[0;34m(sock, keyfile, certfile, cert_reqs, ca_certs, server_hostname, ssl_version, ciphers, ssl_context, ca_cert_dir, key_password, ca_cert_data, tls_in_tls)\u001b[0m\n\u001b[1;32m    442\u001b[0m \u001b[39mif\u001b[39;00m ca_certs \u001b[39mor\u001b[39;00m ca_cert_dir \u001b[39mor\u001b[39;00m ca_cert_data:\n\u001b[1;32m    443\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 444\u001b[0m         context\u001b[39m.\u001b[39;49mload_verify_locations(ca_certs, ca_cert_dir, ca_cert_data)\n\u001b[1;32m    445\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mOSError\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    446\u001b[0m         \u001b[39mraise\u001b[39;00m SSLError(e) \u001b[39mfrom\u001b[39;00m \u001b[39me\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "def main():\n",
    "    import matplotlib.pyplot as plt\n",
    "    from collections import Counter\n",
    "    import numpy as np\n",
    "    import pandas as pd\n",
    "\n",
    "    enzyme_file = \"test_enzyme.fa\"\n",
    "\n",
    "    # Read genome and restriction enzyme sites\n",
    "    genome = fetch_fasta_from_ncbi(\"NC_037328.1\")\n",
    "    enzymes = restriction_sites(enzyme_file)\n",
    "    \n",
    "    # Create a dictionary to store the restriction sites for the selected enzymes\n",
    "    # selected_enzymes = [enzyme.strip() for enzyme in input(\"Enter the names of the enzymes separated by commas (e.g., EcoRI, HindIII): \").split(',')]\n",
    "    selected_enzymes = \"EcoRI, HindIII\".split(',')\n",
    "    # Find restriction s]ites for the selected enzymes and generate the selected_restriction_sites dictionary\n",
    "    from collections import defaultdict\n",
    "    selected_restriction_sites = defaultdict(set) # Use defaultdict to avoid key errors and store unique restriction sites\n",
    "    for enzyme_name in selected_enzymes:\n",
    "        enzyme_name = enzyme_name.strip() # Remove whitespaces\n",
    "        if enzyme_name in enzymes:\n",
    "            enzyme_sites = find_restriction_sites(genome, enzyme_name)\n",
    "            for chr, sites in enzyme_sites.items(): # Add the restriction sites to the selected_restriction_sites dictionary\n",
    "                selected_restriction_sites[chr].update(sites)\n",
    "        else:\n",
    "            print(\"{enzyme} not found!\".format(enzyme=enzyme_name))\n",
    "    \n",
    "    for chr in selected_restriction_sites: # Sort the restriction sites and convert the set to sorted list\n",
    "        selected_restriction_sites[chr] = sorted(list(selected_restriction_sites[chr]))\n",
    "    \n",
    "    # Simulate the fragmentation during genome DNA extraction\n",
    "    distances = []\n",
    "    for chr, sites in selected_restriction_sites.items():\n",
    "        for i in range(500):\n",
    "            sites_with_fragmentation = random_fragment(sites, 1, 4000)\n",
    "            distances_in_chr = [sites_with_fragmentation[i+1] - sites_with_fragmentation[i] for i in range(len(sites_with_fragmentation)-1)]\n",
    "            distances.extend(distances_in_chr)\n",
    "\n",
    "\n",
    "    # Summarize the fragment sizes by printing the sum of the fragment sizes in each bin\n",
    "    bins = [i for i in range(0, 20000, 1)]\n",
    "    \n",
    "    # Make a pandas dataframe to store the fragment sizes and their counts. Using bins of size 1\n",
    "    distances_count = Counter(distances)\n",
    "\n",
    "    # Sum the counts of the fragment sizes in each bin\n",
    "    df = pd.DataFrame.from_dict(distances_count, orient='index').reset_index()\n",
    "    df.columns = ['size', 'count']\n",
    "\n",
    "    # df.to_csv(\"fragment_sizes.csv\", index=False)\n",
    "    df[\"bases\"] = df[\"size\"] * df[\"count\"]\n",
    "    \n",
    "    # Create an index column with the fragment sizes\n",
    "    df.set_index(\"size\", inplace=True)\n",
    "    \n",
    "    # Delete the count column\n",
    "    del df[\"count\"]\n",
    "    \n",
    "    # Convert the df index into bins\n",
    "    df = df.groupby(pd.cut(df.index, bins=bins)).sum()\n",
    "\n",
    "    # Change zeros and NaNto 1 to avoid log(0) error\n",
    "    df = df.replace(0, 1)\n",
    "    df = df.fillna(1)\n",
    "\n",
    "    # Change the index into an integer by converting the string into a list and taking the first element\n",
    "    df.index = [int(str(i)[1:-1].split(\",\")[1]) for i in df.index]\n",
    "\n",
    "    # Normalize the bases column\n",
    "    df[\"bases\"] = df[\"bases\"] / df[\"bases\"].sum() * 1000\n",
    "\n",
    "    # Plot the simulated gel electrophoresis\n",
    "    plt.figure(figsize=(2, 10))\n",
    "    plt.imshow(df, aspect='auto', cmap=\"gray\")\n",
    "    plt.yscale('log')\n",
    "    plt.ylim(50, df.index[-1])\n",
    "\n",
    "    # set yticks from a list\n",
    "    ticks = [50, 100, 200, 300, 400, 500, 600, 700, 800, 900, 1000, 1500, 2000, 2500, 3000, 4000, 5000, 6000, 8000, 10000]\n",
    "    plt.yticks(ticks, ticks)\n",
    "\n",
    "\n",
    "    plt.colorbar(label='DNA Amount (ng per µg)')\n",
    "    xlabel = \" + \".join(selected_enzymes)\n",
    "    plt.xlabel(xlabel)\n",
    "    plt.ylabel('Fragment Size (bp)')\n",
    "    plt.title(\"Simulated DNA Gel after \" + \" + \".join(selected_enzymes) + \" Digestion\")\n",
    "    #plt.savefig(\" + \".join(selected_enzymes) + \" Gel.jpg\", dpi=1000, bbox_inches='tight')\n",
    "    plt.show()\n",
    "\n",
    "    # Make another plot based on df, with index as x and bases as y\n",
    "    from scipy.interpolate import interp1d\n",
    "    import seaborn as sns\n",
    "\n",
    "    bins_new = [i for i in range(0, 20000, 20)]\n",
    "    df_new = df.groupby(pd.cut(df.index, bins=bins_new)).sum()\n",
    "    df_new[\"bases\"] = df_new[\"bases\"]/df_new[\"bases\"].sum() * 1000\n",
    "    df_new.index = [int(str(i)[1:-1].split(\",\")[1]) for i in df_new.index]\n",
    "\n",
    "    plt.figure(figsize=(15, 5))\n",
    "    sns.lineplot(x=df_new.index, y=df_new[\"bases\"])\n",
    "    #plt.xscale('log')\n",
    "    #plt.yscale('log')\n",
    "    plt.xlim(0, df_new.index[-1])\n",
    "    plt.ylim(0, df_new[\"bases\"].max() + 2)\n",
    "    plt.xlabel(\"Fragment Size (bp)\")\n",
    "    plt.ylabel(\"DNA Amount (ng per µg)\")\n",
    "    plt.title(\"Simulated Femto Pulse after \" + \" + \".join(selected_enzymes) + \" Digestion\")\n",
    "    # Annotate the data point with the highest DNA amount\n",
    "    plt.annotate(str(round(df[\"bases\"].idxmax())), xy=(df[\"bases\"].idxmax(), df_new[\"bases\"].max()), xytext=(df[\"bases\"].idxmax(), df_new[\"bases\"].max()*1.05),\n",
    "                 rotation=90, ha='center', va='bottom', color=\"blue\")\n",
    "    # plt.savefig(\" + \".join(selected_enzymes) + \" Femto Pulse.jpg\", dpi=1000)\n",
    "    plt.show()\n",
    "   \n",
    "    return distances # Return the distances list to be used in the plot\n",
    "\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
